From: Antonio Terceiro <terceiro@debian.org>
Date: Wed, 13 Oct 2021 19:57:55 -0300
Subject: rand_init: fix off-by-one error

This was found when debugging a Bus error on mipsel. The problem was
isolated to this test case:

> $ ./miniruby -e 'srand(0)'                     [289/307]
> -e:1: [BUG] Bus Error at 0x00000000
> ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [mipsel-linux-gnu]
>
> -- Control frame information -----------------------------------------------
> c:0003 p:---- s:0011 e:000010 CFUNC  :srand
> c:0002 p:0004 s:0006 e:000005 EVAL   -e:1 [FINISH]
> c:0001 p:0000 s:0003 E:000818 (none) [FINISH]
>
> -- Ruby level backtrace information ----------------------------------------
> -e:1:in `<main>'
> -e:1:in `srand'
>
> -- C level backtrace information -------------------------------------------
> ./miniruby(0x2b6e30) [0x55842e30]
>
> -- Other runtime information -----------------------------------------------
>
> * Loaded script: -e
>
> * Loaded features:
>
>     0 enumerator.so
>     1 thread.rb
>     2 rational.so
>     3 complex.so
>     4 ruby2_keywords.rb
>
> * Process memory map:
>
> 5558c000-558f9000 r-xp 00000000 fe:01 5255053                            /home/terceiro/ruby3.0-3.0.2/miniruby
> 55909000-5590c000 r-xp 0036d000 fe:01 5255053                            /home/terceiro/ruby3.0-3.0.2/miniruby
> 5590c000-5590f000 rwxp 00370000 fe:01 5255053                            /home/terceiro/ruby3.0-3.0.2/miniruby
> 5590f000-55916000 rwxp 00000000 00:00 0
> 55a14000-55af7000 rwxp 00000000 00:00 0                                  [heap]
> 7436f000-757eb000 r-xs 00000000 fe:01 5255053                            /home/terceiro/ruby3.0-3.0.2/miniruby
> 757eb000-75812000 r-xp 00000000 fe:01 5665194                            /lib/mipsel-linux-gnu/libgcc_s.so.1
> 75812000-75822000 ---p 00027000 fe:01 5665194                            /lib/mipsel-linux-gnu/libgcc_s.so.1
> 75822000-75823000 rwxp 00027000 fe:01 5665194                            /lib/mipsel-linux-gnu/libgcc_s.so.1
> 75823000-75824000 ---p 00000000 00:00 0
> 75824000-75875000 rwxp 00000000 00:00 0
> 75875000-75876000 ---p 00000000 00:00 0
> 75876000-758c7000 rwxp 00000000 00:00 0
> 758c7000-758c8000 ---p 00000000 00:00 0
> 758c8000-75919000 rwxp 00000000 00:00 0
> 75919000-7591a000 ---p 00000000 00:00 0
> 7591a000-7596b000 rwxp 00000000 00:00 0
> 7596b000-7596c000 ---p 00000000 00:00 0
> 7596c000-759bd000 rwxp 00000000 00:00 0
> 759bd000-759be000 ---p 00000000 00:00 0
> 759be000-75a0f000 rwxp 00000000 00:00 0
> 75a0f000-75a10000 ---p 00000000 00:00 0
> 75a10000-75a61000 rwxp 00000000 00:00 0
> 75a61000-75a62000 ---p 00000000 00:00 0
> 75a62000-77b6e000 rwxp 00000000 00:00 0
> 77b6e000-77cf5000 r-xp 00000000 fe:01 5656753                            /lib/mipsel-linux-gnu/libc-2.32.so
> 77cf5000-77d05000 ---p 00187000 fe:01 5656753                            /lib/mipsel-linux-gnu/libc-2.32.so
> 77d05000-77d08000 r-xp 00187000 fe:01 5656753                            /lib/mipsel-linux-gnu/libc-2.32.so
> 77d08000-77d0b000 rwxp 0018a000 fe:01 5656753                            /lib/mipsel-linux-gnu/libc-2.32.so
> 77d0b000-77d0d000 rwxp 00000000 00:00 0
> 77d0d000-77d7b000 r-xp 00000000 fe:01 5657223                            /lib/mipsel-linux-gnu/libm-2.32.so
> 77d7b000-77d8a000 ---p 0006e000 fe:01 5657223                            /lib/mipsel-linux-gnu/libm-2.32.so
> 77d8a000-77d8b000 r-xp 0006d000 fe:01 5657223                            /lib/mipsel-linux-gnu/libm-2.32.so
> 77d8b000-77d8c000 rwxp 0006e000 fe:01 5657223                            /lib/mipsel-linux-gnu/libm-2.32.so
> 77d8c000-77dc6000 r-xp 00000000 fe:01 5656825                            /lib/mipsel-linux-gnu/libcrypt.so.1.1.0
> 77dc6000-77dd5000 ---p 0003a000 fe:01 5656825                            /lib/mipsel-linux-gnu/libcrypt.so.1.1.0
> 77dd5000-77dd6000 r-xp 00039000 fe:01 5656825                            /lib/mipsel-linux-gnu/libcrypt.so.1.1.0
> 77dd6000-77dd7000 rwxp 0003a000 fe:01 5656825                            /lib/mipsel-linux-gnu/libcrypt.so.1.1.0
> 77dd7000-77ddf000 rwxp 00000000 00:00 0
> 77ddf000-77de2000 r-xp 00000000 fe:01 5656948                            /lib/mipsel-linux-gnu/libdl-2.32.so
> 77de2000-77df1000 ---p 00003000 fe:01 5656948                            /lib/mipsel-linux-gnu/libdl-2.32.so
> 77df1000-77df2000 r-xp 00002000 fe:01 5656948                            /lib/mipsel-linux-gnu/libdl-2.32.so
> 77df2000-77df3000 rwxp 00003000 fe:01 5656948                            /lib/mipsel-linux-gnu/libdl-2.32.so
> 77df3000-77e6f000 r-xp 00000000 fe:01 5664300                            /usr/lib/mipsel-linux-gnu/libgmp.so.10.4.1
> 77e6f000-77e7e000 ---p 0007c000 fe:01 5664300                            /usr/lib/mipsel-linux-gnu/libgmp.so.10.4.1
> 77e7e000-77e7f000 r-xp 0007b000 fe:01 5664300                            /usr/lib/mipsel-linux-gnu/libgmp.so.10.4.1
> 77e7f000-77e80000 rwxp 0007c000 fe:01 5664300                            /usr/lib/mipsel-linux-gnu/libgmp.so.10.4.1
> 77e80000-77e88000 r-xp 00000000 fe:01 5656667                            /lib/mipsel-linux-gnu/librt-2.32.so
> 77e88000-77e97000 ---p 00008000 fe:01 5656667                            /lib/mipsel-linux-gnu/librt-2.32.so
> 77e97000-77e98000 r-xp 00007000 fe:01 5656667                            /lib/mipsel-linux-gnu/librt-2.32.so
> 77e98000-77e99000 rwxp 00008000 fe:01 5656667                            /lib/mipsel-linux-gnu/librt-2.32.so
> 77e99000-77eb6000 r-xp 00000000 fe:01 5656815                            /lib/mipsel-linux-gnu/libpthread-2.32.so
> 77eb6000-77ec5000 ---p 0001d000 fe:01 5656815                            /lib/mipsel-linux-gnu/libpthread-2.32.so
> 77ec5000-77ec6000 r-xp 0001c000 fe:01 5656815                            /lib/mipsel-linux-gnu/libpthread-2.32.so
> 77ec6000-77ec7000 rwxp 0001d000 fe:01 5656815                            /lib/mipsel-linux-gnu/libpthread-2.32.so
> 77ec7000-77ec9000 rwxp 00000000 00:00 0
> 77ec9000-77ee3000 r-xp 00000000 fe:01 5657005                            /lib/mipsel-linux-gnu/libz.so.1.2.11
> 77ee3000-77ef2000 ---p 0001a000 fe:01 5657005                            /lib/mipsel-linux-gnu/libz.so.1.2.11
> 77ef2000-77ef3000 r-xp 00019000 fe:01 5657005                            /lib/mipsel-linux-gnu/libz.so.1.2.11
> 77ef3000-77ef4000 rwxp 0001a000 fe:01 5657005                            /lib/mipsel-linux-gnu/libz.so.1.2.11
> 77ef4000-77f1b000 r-xp 00000000 fe:01 5656803                            /lib/mipsel-linux-gnu/ld-2.32.so
> 77f20000-77f24000 rwxp 00000000 00:00 0
> 77f28000-77f2a000 rwxp 00000000 00:00 0
> 77f2a000-77f2b000 r-xp 00026000 fe:01 5656803                            /lib/mipsel-linux-gnu/ld-2.32.so
> 77f2b000-77f2c000 rwxp 00027000 fe:01 5656803                            /lib/mipsel-linux-gnu/ld-2.32.so
> 7ef06000-7f705000 rwxp 00000000 00:00 0                                  [stack]
> 7fef5000-7fef6000 r-xp 00000000 00:00 0
> 7ffbf000-7ffc0000 r--p 00000000 00:00 0                                  [vvar]
> 7ffc0000-7ffc1000 r-xp 00000000 00:00 0                                  [vdso]
>
>
> Aborted

This has just been found on mipsel, probably due to how the stack is
managed on that architecture. rand_init_default, right above rand_init,
calls ALLOCV_N with len+1, so make rand_init do the same. This fixes the
issue for me, and causes no negative effects on other architectures.

Signed-off-by: Antonio Terceiro <terceiro@debian.org>
Signed-off-by: Aurelien Jarno <aurel32@debian.org>
---
 random.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/random.c b/random.c
index 7567d13..20c2472 100644
--- a/random.c
+++ b/random.c
@@ -369,7 +369,7 @@ rand_init(const rb_random_interface_t *rng, rb_random_t *rnd, VALUE seed)
     int sign;
 
     len = rb_absint_numwords(seed, 32, NULL);
-    buf = ALLOCV_N(uint32_t, buf0, len);
+    buf = ALLOCV_N(uint32_t, buf0, len+1);
     sign = rb_integer_pack(seed, buf, len, sizeof(uint32_t), 0,
         INTEGER_PACK_LSWORD_FIRST|INTEGER_PACK_NATIVE_BYTE_ORDER);
     if (sign < 0)
